influxql/decoder.go

```diff
@@ -0,0 +1,188 @@
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/flux/array"
+
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/execute"
+	"github.com/influxdata/flux/execute/tablebuilder"
+	"github.com/influxdata/flux/values"
+)
+
+// NewResultDecoder will construct a new result decoder for an influxql response.
+func NewResultDecoder(a *execute.Allocator) flux.MultiResultDecoder {
+	return &resultDecoder{a: a}
+}
+
+type resultDecoder struct {
+	a *execute.Allocator
+}
+
+func (dec *resultDecoder) Decode(r io.ReadCloser) (flux.ResultIterator, error) {
+	var resp Response
+	if err := json.NewDecoder(r).Decode(&resp); err != nil {
+		return nil, err
+	}
+	return &resultIterator{resp: &resp, a: dec.a}, nil
+}
+
+type resultIterator struct {
+	resp *Response
+	a    *execute.Allocator
+}
+
+func (ri *resultIterator) More() bool {
+	return len(ri.resp.Results) > 0 && ri.resp.Results[0].Err == ""
+}
+
+func (ri *resultIterator) Next() flux.Result {
+	res := ri.resp.Results[0]
+	ri.resp.Results = ri.resp.Results[1:]
+	return &result{res: &res, a: ri.a}
+}
+
+func (ri *resultIterator) Cancel() {
+	ri.resp.Results = nil
+}
+
+func (ri *resultIterator) Err() error {
+	if ri.resp.Err != "" {
+		return errors.New(ri.resp.Err)
+	} else if len(ri.resp.Results) > 0 && ri.resp.Results[0].Err != "" {
+		return errors.New(ri.resp.Results[0].Err)
+	}
+	return nil
+}
+
+type result struct {
+	res *Result
+	a   *execute.Allocator
+}
+
+func (r *result) Name() string {
+	return strconv.Itoa(r.res.StatementID)
+}
+
+func (r *result) Tables() flux.TableIterator {
+	return r
+}
+
+func (r *result) Do(f func(tbl flux.Table) error) error {
+	// Iterate through each series within the table.
+	for _, series := range r.res.Series {
+		// Find the time column.
+		// TODO(jsternberg): Allow the time column to be customized.
+		timeCol := -1
+		for i, col := range series.Columns {
+			if col == "time" {
+				timeCol = i
+				break
+			}
+		}
+
+		// Within each series, we have a common group key base. Construct that here.
+		seriesKeyBuilder := execute.NewGroupKeyBuilder(nil)
+		if series.Name != "" {
+			seriesKeyBuilder.AddKeyValue("_measurement", values.NewString(series.Name))
+		}
+
+		tagKeys := make([]string, 0, len(series.Tags))
+		for k := range series.Tags {
+			tagKeys = append(tagKeys, k)
+		}
+		sort.Strings(tagKeys)
+		for _, k := range tagKeys {
+			seriesKeyBuilder.AddKeyValue(k, values.NewString(series.Tags[k]))
+		}
+
+		seriesKey, err := seriesKeyBuilder.Build()
+		if err != nil {
+			return err
+		}
+
+		// Iterate through the columns and create a new table for each field.
+		for i, col := range series.Columns {
+			if i == timeCol {
+				// Skip the time column if one was found.
+				continue
+			}
+
+			// Construct a new table.
+			b := tablebuilder.New(r.a)
+
+			// If we have a time column, do it here.
+			if timeCol >= 0 {
+				if err := b.Times("_time").Do(func(b array.TimeBuilder) error {
```

Comment:
This seems to be exposing the data at too low-level.  

A function author should not be thinking "I need to remember to Reserve space for my column".  This seems like a long way around compared to the existing method: `builder.AppendTime(timeIndex, values.Time(t.UnixNano))`